From 60bc1ab99cbeb8d988f8eab51805e9341681f707 Mon Sep 17 00:00:00 2001
From: Yasushi SHOJI <yashi@spacecubics.com>
Date: Fri, 14 Nov 2025 12:45:29 +0900
Subject: [PATCH] mtd: spi-nor: spansion: use spi_nor_get_params() for size check

The stacked-memories update from Xilinx[1] changed spi_nor->params
from a single pointer to per-flash parameter storage. Accessing
nor->params directly is no longer correct when stacked configuration
is enabled.

Fetch the flash0 parameter block via spi_nor_get_params(nor, 0) and
use it for the SZ_64M check in s25fs_s_nor_smpt_map_id_dummy(),
matching the upstream driver expectations.

After applying S25FS-S fixes[2], this change is need for
xlnx_rebase_v6.12_LTS_2025.1_update.

[1]: https://github.com/Xilinx/linux-xlnx/commit/d5b1382cc533d41ae3e1d652d9032afc701e7892
[2]: https://lists.infradead.org/pipermail/linux-mtd/2025-November/110844.html

Signed-off-by: Yasushi SHOJI <yashi@spacecubics.com>
---
 drivers/mtd/spi-nor/spansion.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/mtd/spi-nor/spansion.c b/drivers/mtd/spi-nor/spansion.c
index 956b8c94bc12..1569d9c63967 100644
--- a/drivers/mtd/spi-nor/spansion.c
+++ b/drivers/mtd/spi-nor/spansion.c
@@ -864,7 +864,8 @@ static void s25fs_s_nor_smpt_map_id_dummy(const struct spi_nor *nor, u8 *map_id)
 	 * setting SMPT by overwriting the CR3NV[1] value to 1, as the table
 	 * expects.
 	 */
-	if (nor->params->size == SZ_64M)
+	struct spi_nor_flash_parameter *params = spi_nor_get_params(nor, 0);
+	if (params->size == SZ_64M)
 		*map_id |= BIT(0);
 }
 
